### DISCUSSION TITLE
Morphing <tr> fails

### DISCUSSION THREAD

**Discussion by Alex-Sichkar** [alpinejs/alpine#3019] [7. Bugs (archived)]
The following case fails:
```html
<table>
  <tr id="tr">
    <td>Text</td>
  </tr>
</table>
```
```js
let el = document.querySelector('#tr')
Alpine.morph(el, '<tr><td>Morphed</td></tr>')
```
with an exception:
```
alpinejs.morph.cdn.min.js:2 Uncaught (in promise) TypeError: Cannot set properties of null (setting '_x_dataStack')
```
Am I doing something wrong?

**Comment by samwillis** [comment]
I also believe this is a bug with the Morph plugin. I have put together a JSBin reproduction of this: https://jsbin.com/duvamixiro/edit?html,output

Although I haven't looked in detail I believe morph works by creating an un-mounted element (probably a div), and settings its `innerHTML` to the new html you would like to morph to. The issue is that `<tr>` is only valid as a direct child of a table. If you try this in your browser console you can see that the `<td><td>` is striped out of the created DOM:

![image](https://user-images.githubusercontent.com/31130/177206351-06a43c72-63cc-4ca5-ac81-8b219c837007.png)

I suspect the solution is to inspect the string prior to creating the element to insert it into, work out what the top level tag is, and use an element that would allow it as a child.

**Reply by samwillis** [reply]
In fact here is the code from Morph.js:

https://github.com/alpinejs/alpine/blob/763e287842da1d9e0a32c52015cfd5c8f33dbac7/packages/morph/src/dom.js#L59-L61

And what happens in the browser:

![image](https://user-images.githubusercontent.com/31130/177207702-27446793-dd32-47be-aac8-401fcb549626.png)

There is some information here on alternative methods: https://stackoverflow.com/questions/43102944/cannot-create-documentfragment-storing-td-tr-or-th

**Reply by samwillis** [reply]
I have submitted a pull request to change the DOM fragment creation method to enable table support: https://github.com/alpinejs/alpine/pull/3020
